### PR TITLE
[hf-callback] General mechanism to track HF Jobs

### PR DESCRIPTION
**Pull request by valayDave**
## Example
```python
# import the decorator compatible with callback
from metaflow.plugins.hf_info_card import huggingface_card

class HFCBFlow(FlowSpec):

    @huggingface_card
    @step
    def start(self):
        my_hf_trainer_fn()
        self.next(self.end)

    @step
    def end(self):
        pass

```

```python
# in your training function drop the callback
from metaflow.plugins.hf_info_card import MetaflowHuggingFaceCardCallback

def my_hf_trainer_fn():
    ...
    ...
    Trainer(
        ...
        callbacks=[MetaflowHuggingFaceCardCallback(
            tracked_metrics=["val_loss", "train_loss"],
        )]

    )
```

### CODE REVIEW

File: examples/metaflow-hf-trainer-card-callback/train.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+try: # in case you want to download to local env without deepspeed
+    from deepspeed import comm as dist
+    from deepspeed import init_distributed
+except ImportError:
+    pass
+
+from metaflow.plugins.hf_info_card import MetaflowHuggingFaceCardCallback
+
+def main(
```
Comment (emattia) [inline examples/metaflow-hf-trainer-card-callback/train.py:26]: Let's add an eval set so we have a more meaningful experiment tracking view on first UX:

```python
def main(
    checkpoint_dir: str = "training_outputs",
    ds_name: str = "rotten_tomatoes",
    model_name: str = "distilbert-base-uncased",
    train_split: str = "train[:50%]",
    eval_split: str = "validation",
    run_id: str = None,
    flow_name: str = None,
    local_rank: int = None,
    global_rank: int = None,
):
    train_ds = load_dataset(ds_name, split=train_split)
    eval_ds = load_dataset(ds_name, split=eval_split)
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForSequenceClassification.from_pretrained(model_name)
    tokenize_dataset = lambda dataset: tokenizer(dataset["text"])
    train_ds = train_ds.map(tokenize_dataset, batched=True)
    eval_ds = eval_ds.map(tokenize_dataset, batched=True)

    data_collator = DataCollatorWithPadding(tokenizer=tokenizer)
    training_args = TrainingArguments(
        output_dir=checkpoint_dir,
        overwrite_output_dir=True,
        learning_rate=1e-5,
        do_train = True,
        do_eval = True,
        evaluation_strategy = "epoch",
        disable_tqdm = True, # this example logs progess bars to Metaflow GUI using MetaflowHuggingFaceCardCallback
        per_device_train_batch_size=8,
        per_device_eval_batch_size=8,
        num_train_epochs=5,
        logging_steps=1,
        # save 2 checkpoints (best one and last one)
        save_total_limit=0,
        save_steps=200,
        save_strategy="steps",
        load_best_model_at_end=False,
        save_on_each_node=False,  # default. Notice relation to DeepspeedHFTrainerS3Sync.push_from_all_nodes.
        warmup_steps=1,
        report_to="none",
    )
    trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=train_ds,
        eval_dataset=eval_ds,
        tokenizer=tokenizer,
        data_collator=data_collator,
        callbacks=[
            MetaflowHuggingFaceCardCallback(
                tracked_metrics=[
                    "loss",
                    "learning_rate",
                    "grad_norm",
                    "eval_loss"
                ]
            )
        ],
    )
    trainer.train()
```

File: metaflow_extensions/deepspeed/plugins/mfextinit_deepspeed.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
 STEP_DECORATORS_DESC = [
     ("deepspeed", ".deepspeed_libs.deepspeed_decorator.DeepspeedDecorator")
 ]
-__mf_promote_submodules__ = ["az_store", "hf_callbacks", "deepspeed_libs",]
\ No newline at end of file
+__mf_promote_submodules__ = ["az_store", "hf_callbacks", "deepspeed_libs", "hf_info_card"]
```
Comment (emattia) [inline metaflow_extensions/deepspeed/plugins/mfextinit_deepspeed.py]: At some point, can clean up HF cards, and move them up into a repo that isn't specific to deepspeed.

File: metaflow_extensions/deepspeed/plugins/hf_info_card.py [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+    x_name="u",
+    y_name="v",
+    xtitle=None,
+    ytitle=None,
+    width=600,
+    height=400,
+    with_params=True,
+    x_axis_temporal=False,
+):
+    parameters = [
```
Comment (emattia) [inline metaflow_extensions/deepspeed/plugins/hf_info_card.py]: Nits
- How to left justify the y axis of vertically stacked Vega charts like is happening in this example. Right now, if the ytick values have many numbers on one plot, it offsets the y-axis for that plot relative to others. 
- How to fix plots so x tick val limits are [0, max number steps]. Similar to above, it is helpful to see everything stacked on same x values. For plots like evals, the first data is at step N, and the first eval loss value on the chart will be N minus a few instead of aligned with the train loss.
- [twinx](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.twinx.html)-like mechanism could be nice for viewing train and eval curves together.

Comment (valayDave) [reply metaflow_extensions/deepspeed/plugins/hf_info_card.py]: help fix it plz :) I have created general purpose machinery for building different types of cards. All we need to do is figure ways to populate it with the right data.
